pam_kwallet.c

```diff
@@ -0,0 +1,823 @@
+        struct stat info;
+        if (stat(path, &info) == 0 &&
+            info.st_size != 0 &&
+            S_ISREG(info.st_mode)) {
+            exit(0);
+        }
+
+        unlink(path);//in case the file already exists
+
+        char *dir = strdup(path);
+        dir[strlen(dir) - 14] = '\0';//remove kdewallet.salt
+        mkpath(dir); //create the path in case it does not exists
+        free(dir);
+
+        char *salt = gcry_random_bytes(KWALLET_PAM_SALTSIZE, GCRY_STRONG_RANDOM);
+        const int fd = open(path, O_CREAT | O_WRONLY | O_TRUNC | O_CLOEXEC, 0600);
+
+        //If the file can't be created
+        if (fd == -1) {
+            syslog(LOG_ERR, "%s: Couldn't open file: %s because: %d-%s", logPrefix, path, errno, strerror(errno));
+            exit(-2);
+        }
+
+        const ssize_t wlen = write(fd, salt, KWALLET_PAM_SALTSIZE);
+        close(fd);
+        if (wlen != KWALLET_PAM_SALTSIZE) {
+            syslog(LOG_ERR, "%s: Short write to file: %s", logPrefix, path);
+            unlink(path);
+            exit(-2);
+        }
+
+        exit(0); // success
+    } else {
+        // pam process, just wait for child to finish
+        int status;
+        waitpid(pid, &status, 0);
+        if (status != 0) {
+            pam_syslog(pamh, LOG_ERR, "%s: Couldn't create salt file", logPrefix);
+        }
+    }
+}
+
+static int readSaltFile(pam_handle_t *pamh, char *path, struct passwd *userInfo, char *saltOut)
+{
+    int readSaltPipe[2];
+    if (pipe(readSaltPipe) < 0) {
+        pam_syslog(pamh, LOG_ERR, "%s: Couldn't create read salt pipes", logPrefix);
+        return 0;
+    }
+
+    const pid_t pid = fork();
+    if (pid == -1) {
+        syslog(LOG_ERR, "%s: Couldn't fork to read salt file", logPrefix);
+        close(readSaltPipe[0]);
+        close(readSaltPipe[1]);
+        return 0;
+    } else if (pid == 0) {
+        // Child process
+        close(readSaltPipe[0]); // we won't be reading from the pipe
+        if (drop_privileges(userInfo) < 0) {
+            syslog(LOG_ERR, "%s: could not set gid/uid/euid/egit for salt file reading", logPrefix);
+            free(path);
+            close(readSaltPipe[1]);
+            exit(-1);
+        }
+
+        struct stat info;
+        if (stat(path, &info) != 0 || info.st_size == 0 || !S_ISREG(info.st_mode)) {
+            syslog(LOG_ERR, "%s: Failed to ensure %s looks like a salt file", logPrefix, path);
+            free(path);
+            close(readSaltPipe[1]);
+            exit(-1);
+        }
+
+        const int fd = open(path, O_RDONLY | O_CLOEXEC);
+        if (fd == -1) {
+            syslog(LOG_ERR, "%s: Couldn't open file: %s because: %d-%s", logPrefix, path, errno, strerror(errno));
+            free(path);
+            close(readSaltPipe[1]);
+            exit(-1);
+        }
+        free(path);
+        char salt[KWALLET_PAM_SALTSIZE] = {};
+        const ssize_t bytesRead = read(fd, salt, KWALLET_PAM_SALTSIZE);
```

Comment:
Here as well as in other spots a partial `read()` is considered an error. A partial read is perfectly normal, though, and would rather require something parallel to what `better_write()` already does for writing.